### DISCUSSION TITLE
Anonymous wildcard doesn't seem to work: _

### DISCUSSION THREAD

**Discussion by emiasims** [tree-sitter/tree-sitter#1361] [Q&A] [answered]
If I have a node that consists only of anonymous wildcards (here: `(item)`), and want to capture the first one that matches some keyword, I would expect it to look something like:
`(item . _ @keyword(#eq? @keyword "TODO"))`

This expectation is from
> There are two types, (_) will match any named node, and _ will match any named or anonymous node

In the [query syntax docs](https://tree-sitter.github.io/tree-sitter/using-parsers#query-syntax)

However, that simply doesn't work. The query fails with

> Query compilation failed: QueryError { row: 0, column: 8, offset: 8, message: `(item . _ @keyword (#eq? @keyword \"TODO\"))\n        ^`, kind: Structure }

[The grammar in question](https://github.com/milisims/tree-sitter-org/tree/41eac56ba95cde9326370e862e96d206222a0ea0).

Is this working incorrectly or do I misunderstand the function of the wildcard?

**Comment by maxbrunsfeld** [comment]
>  `item: $ => seq(repeat1(choice($._text, /[ \t]:/)))`

In your grammar, an `item` node never contains any visible children. The `_text` token is hidden because of the leading underscore, and the other token is hidden because it's a regex. Your pattern is trying to match a node *inside* of an `item`. The query compilation error (`kind: Structure`) is telling you that this pattern can't possibly match.

**Reply by emiasims** [reply]
Okay so nodes that are text like `'('` or `'#+'` are 'anonymous' nodes then, and hidden nodes are any unnamed regex nodes or named nodes that start with an underscore? And queries can never match a hidden node, which is different from an anonymous node. Got it, I think.

Thank you!

**Comment by ahlinc** [comment, marked as answer]
@milisims By looking to your grammar I don't see that you have anonymous nodes in your grammar for the `item` rule:
_Excerpts from the grammar:_
```js
    item: $ => seq(repeat1(choice($._text, /[ \t]:/))),

    _text: _ => choice(
      /\p{L}+/,                   // Letters
      /\p{N}+/,                   // Numbers
      /[^\p{Z}\p{L}\p{N}\n\r]/,   // Everything else, minus whitespace
    ),
```
All terminals in the `$._text` are specified by regexps and such terminals are not anonymous nodes such nodes are a third type of nodes that currently don't accessible through tree-sitter API and queries. Documentation also don't accent on this feature properly. Such type of nodes mostly useful to define a part of named nodes or to define named terminals that are specified by rules. For more info see this comment https://github.com/tree-sitter/tree-sitter/issues/1156#issuecomment-877848438 and its issue.
As a workaround you can assign a named rule:
```js
item: $ => seq(repeat1(alias(choice($._text, /[ \t]:/), $.word))
```
Or even anonymous _(but I'm not sure that this will look clear)_:
```js
item: $ => seq(repeat1(alias(choice($._text, /[ \t]:/), 'word'))
```
<details><summary><i>Screenshot:</i></summary>

![image](https://user-images.githubusercontent.com/14666676/130863980-31faab5c-277a-4bfd-bf81-982d5da9bdec.png)

</details>

For the last option your query would work as you expected.

**Reply by emiasims** [reply]
Thanks, this is really helpful. I didn't understand how alias worked with anonymous nodes, and is perfect for this case.
